File: peticiones/datos.py
import requests
import random
from datetime import datetime, timedelta
import time
import sys

# Configuración
API_BASE = "http://localhost:3009/api/dsm44"
TOTAL_EMPLEADOS = 300  # Mantenemos reducido para pruebas
YEAR = 2025
EMPLEADOS = []

AREAS = ["OFICINA", "PRODUCCION", "INVENTARIO"]
TURNOS = ["MATUTINO", "VESPERTINO", "NOCTURNO", "MIXTO"]
STATUS_ASISTENCIA = ["EN_TURNO", "FINALIZADO"]

NOMBRES = [
    "Juan", "Carlos", "Luis", "Miguel", "Jose", "Jorge", "Felipe", "Hector",
    "Marco", "Ricardo", "Fernando", "Pablo", "Rafael", "Alberto", "Andres",
    "Roberto", "Eduardo", "Cristian", "Mario", "Diego", "Omar", "Sergio",
    "Francisco", "Adrian", "Hernan", "Erick", "Kevin", "Oscar", "Manuel",
    "Víctor", "Alan", "Emilio", "Ramiro", "Leonardo", "Esteban", "Bruno",
    "Mauricio", "Gustavo", "Elías", "Tomás",
]

APELLIDOS = [
    "Hernandez", "Martinez", "Gomez", "Perez", "Lopez", "Garcia",
    "Rodriguez", "Sanchez", "Ramirez", "Cruz", "Torres", "Rivera",
    "Gonzalez", "Flores", "Vargas", "Castillo", "Ortega", "Ruiz",
    "Aguilar", "Chavez", "Dominguez", "Silva", "Navarro", "Salazar",
    "Mendoza", "Ponce", "Morales", "Soto", "Camacho", "Cortés",
    "Arias", "Palacios", "Estrada", "Valdez", "Montoya", "Ramos"
]

def random_choice(arr):
    return random.choice(arr)

def generar_nombre_completo():
    return {
        "nombre": random_choice(NOMBRES),
        "apellido_p": random_choice(APELLIDOS),
        "apellido_m": random_choice(APELLIDOS)
    }

def generar_fechas_2025():
    fechas = []
    start_date = datetime(YEAR, 1, 1)
    for i in range(365):
        current_date = start_date + timedelta(days=i)
        fechas.append(current_date.strftime("%Y-%m-%d"))
    return fechas

FECHAS_2025 = generar_fechas_2025()

def make_iso(fecha, hour, minute):
    fecha_obj = datetime.strptime(fecha, "%Y-%m-%d")
    return fecha_obj.replace(hour=hour, minute=minute, second=0).isoformat()

def safe_post(url, data, max_intentos=5, delay_ms=1000):
    intento = 0
    
    while intento < max_intentos:
        try:
            response = requests.post(url, json=data, timeout=30)
            response.raise_for_status()
            return response
        except requests.exceptions.RequestException as e:
            intento += 1
            if intento == max_intentos:
                print(f"❌ ERROR después de {max_intentos} intentos -> {url}")
                print(f"   Mensaje: {str(e)}")
                if hasattr(e, 'response') and e.response is not None:
                    print(f"   Status Code: {e.response.status_code}")
                    print(f"   Response: {e.response.text}")
                return None
            print(f"⚠️  Intento {intento}/{max_intentos} falló -> {url}. Reintentando...")
            time.sleep(delay_ms / 1000)

def register_a_p(fecha, empleado):
    entrada_iso = make_iso(fecha, 8, 0)
    salida_iso = make_iso(fecha, 17, 0)
    
    fecha_entrada = datetime.fromisoformat(entrada_iso)
    fecha_salida = datetime.fromisoformat(salida_iso)
    diferencia_horas = (fecha_salida - fecha_entrada).total_seconds() / 3600
    
    print(f"  📅 Registrando asistencia y producción para empleado {empleado} en {fecha}")
    
    # Registrar asistencia
    asistencia_data = {
        "fecha": fecha,
        "horaEntrada": entrada_iso,
        "horaSalida": salida_iso,
        "entrada": entrada_iso,
        "salida": salida_iso,
        "status": random_choice(STATUS_ASISTENCIA),
        "empleado": empleado,
        "horasTrabajadas": diferencia_horas,
    }
    
    register_asistencia = safe_post(f"{API_BASE}/empleados/create-asistencia", asistencia_data)
    
    # Registrar producción
    produccion_data = {
        "fecha": fecha,
        "turno": random_choice(TURNOS),
        "unidadesProducidas": random.randint(0, 3000),
        "empleado": empleado
    }
    
    register_produccion = safe_post(f"{API_BASE}/empleados/create-produccion", produccion_data)
    
    # Verificar resultados
    resultado = {
        "register_asistencia": register_asistencia,
        "register_produccion": register_produccion
    }
    
    if register_asistencia and register_produccion:
        try:
            asistencia_json = register_asistencia.json()
            produccion_json = register_produccion.json()
            print(f"  ✅ Asistencia creada: {asistencia_json.get('id_reg_a', 'N/A')}")
            print(f"  ✅ Producción creada: {produccion_json.get('id_reg_p', 'N/A')}")
        except Exception as e:
            print(f"  ❌ Error parseando respuesta: {e}")
    else:
        if not register_asistencia:
            print(f"  ❌ Falló asistencia para empleado {empleado}")
        if not register_produccion:
            print(f"  ❌ Falló producción para empleado {empleado}")
    
    return resultado

def main():
    print("🚀 INICIANDO POBLACIÓN DE DATOS...")
    print(f"📡 Conectando a: {API_BASE}")
    
    # Verificar que la API esté disponible
    try:
        test_response = requests.get(f"{API_BASE}/empleados", timeout=10)
        print(f"✅ API conectada correctamente (Status: {test_response.status_code})")
    except Exception as e:
        print(f"❌ NO SE PUEDE CONECTAR A LA API: {e}")
        print("   Asegúrate de que el servidor esté ejecutándose en http://localhost:3000")
        return
    
    # Crear empleados
    print(f"\n👥 CREANDO {TOTAL_EMPLEADOS} EMPLEADOS...")
    empleados_creados = 0
    
    for i in range(1, TOTAL_EMPLEADOS + 1):
        data_nombre = generar_nombre_completo()
        
        empleado_payload = {
            "nombre": data_nombre["nombre"],
            "apellido_p": data_nombre["apellido_p"],
            "apellido_m": data_nombre["apellido_m"],
            "area": random_choice(AREAS),
            "turno": random_choice(TURNOS),
            "salarioDiario": round(200 + random.random() * 350, 2),
            "activo": True
        }
        
        print(f"  Creando empleado {i}/{TOTAL_EMPLEADOS}: {empleado_payload['nombre']} {empleado_payload['apellido_p']}")
        
        register = safe_post(f"{API_BASE}/empleados", empleado_payload)
        
        if register and register.status_code in [200, 201]:
            try:
                data = register.json()
                empleado_id = data.get("id_empleado")
                if empleado_id:
                    print(f"  ✅ Empleado creado: ID {empleado_id}")
                    EMPLEADOS.append(empleado_id)
                    empleados_creados += 1
                else:
                    print(f"  ❌ No se recibió ID del empleado")
            except Exception as e:
                print(f"  ❌ Error procesando respuesta: {e}")
        else:
            print(f"  ❌ Falló la creación del empleado {i}")
    
    if not EMPLEADOS:
        print("❌ NO SE CREARON EMPLEADOS. Saliendo...")
        return
    
    print(f"\n✅ SE CREARON {empleados_creados} EMPLEADOS")
    print("📊 INICIANDO REGISTRO DE ASISTENCIAS Y PRODUCCIÓN...")
    
    # Registrar asistencias y producciones - solo algunos días para prueba
    dias_prueba = FECHAS_2025[:7]  # Solo primera semana para prueba
    total_registros = len(dias_prueba) * len(EMPLEADOS)
    registro_actual = 0
    
    print(f"📅 Registrando {len(dias_prueba)} días para {len(EMPLEADOS)} empleados")
    print(f"📈 Total de registros a crear: {total_registros}")
    
    for fecha in dias_prueba:
        print(f"\n🗓️  FECHA: {fecha}")
        for empleado in EMPLEADOS:
            registro_actual += 1
            print(f"  [{registro_actual}/{total_registros}] ", end="")
            register_a_p(fecha, empleado)
            
            # Pequeña pausa para no saturar el servidor
            time.sleep(0.1)
    
    print(f"\n🎉 PROCESO COMPLETADO!")
    print(f"   • Empleados creados: {len(EMPLEADOS)}")
    print(f"   • Días procesados: {len(dias_prueba)}")
    print(f"   • Registros totales: {registro_actual}")

if __name__ == "__main__":
    main()

File: datos/datos.js
const axios = require("axios");

const API_BASE = "http://localhost:3000/api/dsm44/empleados";
const TOTAL_EMPLEADOS = 80;
const YEAR = 2025;
const EMPLEADOS = [];

const AREAS = ["OFICINA", "PRODUCCION", "INVENTARIO"];
const TURNOS = ["MATUTINO", "VESPERTINO", "NOCTURNO", "MIXTO"];
const STATUS_ASISTENCIA = ["EN_TURNO", "FINALIZADO"];

const NOMBRES = [
  "Juan","Carlos","Luis","Miguel","Jose","Jorge","Felipe","Hector",
  "Marco","Ricardo","Fernando","Pablo","Rafael","Alberto","Andres",
  "Roberto","Eduardo","Cristian","Mario","Diego","Omar","Sergio",
  "Francisco","Adrian","Hernan","Erick","Kevin","Oscar","Manuel",
  "Víctor","Alan","Emilio","Ramiro","Leonardo","Esteban","Bruno",
  "Mauricio","Gustavo","Elías","Tomás",
];

const APELLIDOS = [
  "Hernandez","Martinez","Gomez","Perez","Lopez","Garcia",
  "Rodriguez","Sanchez","Ramirez","Cruz","Torres","Rivera",
  "Gonzalez","Flores","Vargas","Castillo","Ortega","Ruiz",
  "Aguilar","Chavez","Dominguez","Silva","Navarro","Salazar",
  "Mendoza","Ponce","Morales","Soto","Camacho","Cortés",
  "Arias","Palacios","Estrada","Valdez","Montoya","Ramos"
];

const random = (arr) => arr[Math.floor(Math.random() * arr.length)];

const generarNombreCompleto = () => {
    return {
        nombre: random(NOMBRES),
        apellido_p: random(APELLIDOS),
        apellido_m: random(APELLIDOS)
    }
}

const generarFechas2025 = () => {
  const fechas = [];
  const start = new Date(`${YEAR}-01-01T00:00:00Z`);
  for (let d = 0; d < 365; d++) {
    const cur = new Date(start);
    cur.setUTCDate(cur.getUTCDate() + d);
    // Formato DD/MM/YYYY para la API corregida
    const day = String(cur.getUTCDate()).padStart(2, '0');
    const month = String(cur.getUTCMonth() + 1).padStart(2, '0');
    const year = cur.getUTCFullYear();
    fechas.push(`${day}/${month}/${year}`);
  }
  return fechas;
}

const FECHAS_2025 = generarFechas2025();

// Convertir fecha DD/MM/YYYY a Date object
const parseFecha = (fechaStr) => {
  const [day, month, year] = fechaStr.split("/").map(Number);
  return new Date(year, month - 1, day);
}

const safePost = async (url, body, maxIntentos = 10, delayMs = 1000) => {
    let intento = 0;

    while (intento < maxIntentos) {
        try {
            console.log(`Enviando POST a ${url}...`);
            const response = await axios.post(url, body);
            console.log(`✅ POST exitoso a ${url}`);
            return response;
        } catch (e) {
            intento++;
            if (e.response) {
                console.error(`Error ${e.response.status} en ${url}:`, e.response.data);
            } else {
                console.error(`Error de red en ${url}:`, e.message);
            }
            
            if (intento === maxIntentos) {
                console.error(`❌ Fallo Final después de ${maxIntentos} intentos -> ${url}`);
                return null;
            }
            console.warn(`⚠️ Reintento ${intento}/${maxIntentos} -> ${url} en ${delayMs}ms...`);
            await new Promise(resolve => setTimeout(resolve, delayMs));
        }
    }
};

const registerA_P = async (fecha, empleado) => {
    try {
        // Parsear fecha DD/MM/YYYY para cálculos
        const fechaDate = parseFecha(fecha);
        
        // Generar horas aleatorias según el turno del empleado
        const horaEntrada = Math.floor(Math.random() * 3) + 6; // Entre 6:00 y 9:00
        const horaSalida = horaEntrada + 8 + Math.floor(Math.random() * 2); // 8-9 horas después
        
        // Crear fechas completas con horas
        const entradaDate = new Date(fechaDate);
        entradaDate.setHours(horaEntrada, Math.floor(Math.random() * 60), 0, 0);
        
        const salidaDate = new Date(fechaDate);
        salidaDate.setHours(horaSalida, Math.floor(Math.random() * 60), 0, 0);
        
        // Calcular horas trabajadas
        const diferenciaEnMilisegundos = salidaDate - entradaDate;
        const diferenciaEnHoras = parseFloat((diferenciaEnMilisegundos / 3600000).toFixed(2));
        
        // Determinar puntualidad (antes de las 8:00)
        const puntual = horaEntrada < 8 || (horaEntrada === 8 && entradaDate.getMinutes() === 0);
        
        // Formato de fechas para la API (ISO string)
        const fechaISO = fechaDate.toISOString().split('T')[0];
        const entradaISO = entradaDate.toISOString();
        const salidaISO = salidaDate.toISOString();
        
        // Elegir un turno aleatorio
        const turno = random(TURNOS);
        
        // Elegir estatus aleatorio (80% FINALIZADO, 20% EN_TURNO)
        const estatus = Math.random() > 0.2 ? "FINALIZADO" : "EN_TURNO";
        
        console.log(`Registrando asistencia para empleado ${empleado} - Fecha: ${fecha}`);
        
        const register_asistencia = await safePost(`${API_BASE}/create-asistencia`, {
            id_empleado: empleado,
            fecha: fechaISO, // Formato YYYY-MM-DD
            horaEntrada: entradaISO,
            horaSalida: salidaISO,
            horasTrabajadas: diferenciaEnHoras,
            puntual: puntual, // Booleano real
            turno: turno,
            estatus: estatus
        });
        
        console.log(`Registrando producción para empleado ${empleado} - Fecha: ${fecha}`);
        
        const register_produccion = await safePost(`${API_BASE}/create-produccion`, {
            id_empleado: empleado,
            fecha: fechaISO, // Formato YYYY-MM-DD
            turno: turno,
            unidadesProducidas: Math.floor(Math.random() * 300) + 50 // Entre 50 y 350 unidades
        });
        
        return {
            register_asistencia,
            register_produccion
        };
    } catch (error) {
        console.error(`Error en registerA_P para empleado ${empleado}, fecha ${fecha}:`, error.message);
        return null;
    }
}

const main = async() => {
    console.log("🚀 Iniciando inserción de datos...");
    console.log("=".repeat(50));
    
    // Paso 1: Crear empleados
    console.log(`📝 Creando ${TOTAL_EMPLEADOS} empleados...`);
    for (let i = 1; i <= TOTAL_EMPLEADOS; i++) {
        try {
            const dataNombre = generarNombreCompleto();
            
            const empleadoPayload = {
                nombre: dataNombre.nombre,
                apellido_p: dataNombre.apellido_p,
                apellido_m: dataNombre.apellido_m,
                area: random(AREAS),
                turno: random(TURNOS),
                salarioDiario: parseFloat((200 + Math.random() * 350).toFixed(2)),
                activo: true
            };
            
            console.log(`Creando empleado ${i}/${TOTAL_EMPLEADOS}: ${empleadoPayload.nombre} ${empleadoPayload.apellido_p}`);
            
            const register = await safePost(API_BASE, empleadoPayload);
            
            if (register && register.data) {
                const data = register.data;
                console.log(`✅ Empleado creado: ID ${data.id_empleado} - ${data.nombre} ${data.apellido_p}`);
                EMPLEADOS.push(data.id_empleado);
                
                // Pequeña pausa entre empleados para evitar sobrecarga
                await new Promise(resolve => setTimeout(resolve, 100));
            }
        } catch (error) {
            console.error(`Error creando empleado ${i}:`, error.message);
        }
    }
    
    console.log(`✅ ${EMPLEADOS.length} empleados creados exitosamente`);
    console.log("=".repeat(50));
    
    if (EMPLEADOS.length === 0) {
        console.log("❌ No se crearon empleados. Saliendo...");
        return;
    }
    
    // Paso 2: Crear registros de asistencia y producción
    console.log("📊 Creando registros de asistencia y producción...");
    
    // Limitar a algunos días para no sobrecargar (opcional)
    const diasAMostrar = FECHAS_2025.slice(0, 30); // Solo primeros 30 días
    
    for (const fecha of diasAMostrar) {
        console.log(`📅 Procesando fecha: ${fecha}`);
        
        for (const empleado of EMPLEADOS) {
            try {
                // Solo crear registros algunos días (70% de probabilidad)
                if (Math.random() < 0.7) {
                    const registros = await registerA_P(fecha, empleado);
                    
                    if (registros?.register_asistencia && registros?.register_produccion) {
                        console.log(`✅ Registros creados para empleado ${empleado} en ${fecha}`);
                    }
                    
                    // Pequeña pausa entre registros
                    await new Promise(resolve => setTimeout(resolve, 50));
                }
            } catch (error) {
                console.error(`Error procesando empleado ${empleado} en fecha ${fecha}:`, error.message);
            }
        }
    }
    
    console.log("=".repeat(50));
    console.log("🎉 Proceso completado exitosamente!");
    console.log(`📊 Empleados creados: ${EMPLEADOS.length}`);
    console.log(`📅 Días procesados: ${diasAMostrar.length}`);
}

// Manejo de errores global
process.on('unhandledRejection', (error) => {
    console.error('❌ Error no manejado:', error);
});

// Ejecutar el script
main().catch(error => {
    console.error('❌ Error en la ejecución principal:', error);
});